sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc

```diff
@@ -110,157 +145,204 @@ enum class layout {
-
-1- The main compilation mode of SYCL is JIT compilation and partitioning among WIs is implementation defined.
-
-2- Sub group size is not generally fixed.
-
-The code listing below shows a synopsis of these new APIs.
+IMPORTANT: In the current implementation, only the `sub_group` scope
+is supported.
+
+==== Element-Wise Operations
+Besides matrix multiply and add, this extension aims to make it
+possible to perform piece-wise operations on matrices in a SPMD
+manner. `joint_matrix_apply` function performs an element-wise
+operation where the same operation is performed on every element of
+the joint matrix, such that the operation can be performed without knowledge
+of the position of the element within the matrix. Activation functions
+or adding a constant value to every element of the matrix are two
+examples of this usage. When the operation depends on the element
+index of the matrix, an Intel-specific extension is available as part
+of the * link:sycl_ext_intel_matrix.asciidoc[sycl_ext_intel_matrix]
+
+Besides the `Group` and the `joint_matrix` argument,
+`joint_matrix_apply` takes a lambda expression as an argument that
+specifies the specific operation on each of the elements of the input
+matrix.
 
 ```c++
 namespace sycl::ext::oneapi::experimental::matrix {
-   wi_data<group, T, Use, Rows, Cols, Layout> get_wi_data(Group sg, joint_matrix<Group, T, Use, Rows, Cols, Layout> Mat);
-
-template <typename T, size_t Rows, size_t Cols, use Use, layout Layout, typename Group>
-class wi_data {
-  size_t length();
-  wi_element<T, NumRows, NumCols, Use, Layout, Group> operator[](size_t i);
-};
-template <typename T, size_t Rows, size_t Cols,
-          use Use, layout Layout,
-          typename Group = sycl::sub_group>
-class wi_element {
-  operator T();
-  wi_element &operator=(const T &rhs);
-…
-};
+  template<typename Group, typename T, use Use, size_t M, size_t N,
+  layout Layout, typename F>
+  void joint_matrix_apply(Group g, joint_matrix<Group, T, Use, M, N,
+  Layout>C, F&& lambda);
 }
 ```
 
-In the following example `wi_data_c` is a reference to the WI owned portion of the joint matrix `matC`. As such `wi_data_c[i] OP rhs` updates the corresponding matrix element in the joint_matrix `matC`.
-Vectorization along the sub group dimension will get enabled automatically to vectorize the contiguous portion of the matrix. 
-
+In the following example, every element of the matrix `C` is
+multiplied by `alpha`. Then, an activation function, `relu` in this
+example, is applied on each of the elements of `C`.
 
 ```c++
-auto wi_data_c = get_wi_data(sg, matC);
-for (int i = 0; i < wi_data_c.length(); i++)
-        wi_data_c[i] *= alpha;    // Note that the indexing here "i" is in the vector owned by a WI, not in the matrix C        
+joint_matrix_apply(sg, C, [=](T x) {
+    x *= alpha;
+    relu(x);
+});
 ```
+IMPORTANT: `joint_matrix_apply` is not implemented yet.
+
+=== Joint Matrix Additional Types
```

Comment:
> The title of this section should be something like "Support for TF32 floating point" since that is the only other 

Sounds good
 
> * It seems like the purpose of this support is only to reduce the precision of the matrix MAD operation.  As far as the application is concerned, the values appear to be `float`.  When you load values into the matrix, they are `float`.  When you store values out of the matrix, they are `float`.  I presume the per-element access APIs also return `float` values.  Is this a correct understanding?  If so, it would be good to make this clearer in the description somehow.

I say that in the last paragraph but I will clarify more.

> * I assume the application declares a TF32 joint matrix by using the `precision::tf32` type for the `T` template parameter, correct?  The text in this section doesn't actually say that, but it seems like that must be the intent.  Again, it would be good to state this more clearly.  Maybe an example would help too.

Will add that, thanks!

> * I assume there must be overloads for all the joint matrix functions for the `precision::tf32` type, which are different from the ones documented above.  For example, I presume there must be an overload of `joint_matrix_store` like this:
>   If this is the intent, all these overloads should be listed in this section.

Is this really necessary?
In the [actual implementation](https://github.com/intel/llvm/pull/8151/), I did not use overloads per say, but I used traits as follows:
template <typename T> struct helper_traits {
  using element_type = T;
  using storage_element_type = T;
  using fill_argument_type = T;
};

template <> struct helper_traits<precision::tf32> {
  using element_type = precision::tf32;
  using storage_element_type = float;
  using fill_argument_type = float;
};

So in the API, I use storage_element_type and element_type instead of overloads.
In this case, can we just say in the text that  the functions load, store, fill, mad, apply work on joint matrix of type tf32 as well?

> * When the application calls `joint_matrix_load`, I assume the implementation implicitly rounds the `float` value to TF32.  Correct?  What rounding mode is used for this operation?  Round toward zero, round up, round down, etc.?  I presume this should be specified.

That is not true: no rounding is implicitly happening.  As we say before, everything is manipulating floats, except joint_matrix_mad. If users call joint_matrix_mad with no explicit rounding, truncation will happen (only the 19 bits will be taken into account). I will add a clarification about this in the text.

> * When that application calls `joint_matrix_apply`, I presume the callback function receives a value of type `float &`?  This should be clarified.  If the callback function modifies the value, I presume the implementation implicitly rounds the `float` value to TF32 before storing it back to the matrix.  Again, clarify which rounding mode is used.

Same applies to joint_matrix_apply, input joint matrix is tf32, but actual data are floats (storage). no implicit rounding is happening.  

> * Since it seems like there are all these implicit rounding operations that occur, when would the application need to use `round_to_tf32`?  If there is no compelling use case, can we remove that function?

No implicit rounding is happening. data is float.  If users want data to be actually tf32, they have to use the rounding function.